### PR TITLE
Bg haka huta OK

### PR DESCRIPTION
**Pull request by jpburnett**
bg_haka_huta, the coffins used in the bottom of the well. It has been partially documented, got stumped on naming some things. There are a couple of iffy spots in the code. First is in BgHakaHuta_Init, there is an odd *DynaPolyActor new_var used that could be fake. Second is the massive `Actor_Spawn()`s in BgHakaHuta_SpawnEnemies. I don't know if they were coded like that to get the enemies to be in that specific spot in the coffins, or what, but I couldn't figure out anything to clean it up.

### CODE REVIEW

File: src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c [hunk 1]
```diff
@@ -31,25 +37,195 @@ const ActorInit Bg_Haka_Huta_InitVars = {
...
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Haka_Huta/BgHakaHuta_Draw.s")
+void BgHakaHuta_Init(Actor* thisx, GlobalContext* globalCtx) {
+    BgHakaHuta* this = THIS;
+    DynaPolyActor* new_var;
+    CollisionHeader* colHeader;
+
+    // Likely fake, but only way to make a match
+    new_var = &this->dyna;
```
Comment (petrie911) [inline src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c:51]: Did you try using `thisx` and/or `globalCtx2`?

File: src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c [hunk 2]
```diff
@@ -31,25 +37,195 @@ const ActorInit Bg_Haka_Huta_InitVars = {
...
+    colHeader = 0;
+    Actor_ProcessInitChain(&this->dyna.actor, sInitChain);
+    DynaPolyActor_Init(&this->dyna, 1);
+    CollisionHeader_GetVirtual(&D_06000870, &colHeader);
+    this->dyna.bgId = DynaPoly_SetBgActor(globalCtx, &globalCtx->colCtx.dyna, &this->dyna.actor, colHeader);
+    this->unk_16A = ((*new_var).actor.params >> 8) & 0xFF;
+    this->dyna.actor.params &= 0xFF;
+    if (Flags_GetSwitch(globalCtx, this->dyna.actor.params) != 0) {
```
Comment (petrie911) [inline src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c:59]:
```suggestion
    if (Flags_GetSwitch(globalCtx, this->dyna.actor.params)) {
```
We treat Flags_GetSwitch as a boolean.

File: src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c [hunk 3]
```diff
@@ -31,25 +37,195 @@ const ActorInit Bg_Haka_Huta_InitVars = {
...
+    effectVel.z = 0.0f;
+    effectVel.x = -0.5f * phi_f20;
+    effectPos.y = this->dyna.actor.world.pos.y;
+    effectPos.z = this->dyna.actor.world.pos.z;
+    new_Xpos = 50 - ((this->dyna.actor.world.pos.x - this->dyna.actor.home.pos.x) * phi_f20);
+    xPosOffset = new_Xpos * phi_f20;
+
+    for (i = 0; i < 4; i++) {
+        if (i == 2) {
+            effectPos.z += (120.0f * phi_f20);
```
Comment (petrie911) [inline src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c:95]:
```suggestion
            effectPos.z +=120.0f * phi_f20;
```

File: src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c [hunk 4]
```diff
@@ -31,25 +37,195 @@ const ActorInit Bg_Haka_Huta_InitVars = {
...
+        }
+        effectPos.x = this->dyna.actor.home.pos.x - (Rand_ZeroOne() * xPosOffset);
+        scale = ((Rand_ZeroOne() * 10.0f) + 50.0f);
+        func_8002829C(globalCtx, &effectPos, &effectVel, sEffectAccel, &primColor, &envColor, scale, 0xA);
+    }
+}
+
+void func_8087D268(BgHakaHuta* this, GlobalContext* globalCtx, u16 sfx) {
+    Vec3f pos;
```
Comment (petrie911) [inline src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c:104]:
```suggestion
    Vec3f pos;
    
```

File: src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c [hunk 5]
```diff
@@ -31,25 +37,195 @@ const ActorInit Bg_Haka_Huta_InitVars = {
...
+    f32 posOffset;
+
+    if (this->counter != 0) {
+        this->counter--;
+    }
+    if (this->dyna.actor.world.rot.y == 0) {
+        posOffset = 4.0f;
+    } else {
+        posOffset = -4.0f;
+    }
```
Comment (petrie911) [inline src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c:159]: possible ternary?

File: src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c [hunk 6]
```diff
@@ -31,25 +37,195 @@ const ActorInit Bg_Haka_Huta_InitVars = {
...
+    f32 posOffset;
+
+    if (this->counter != 0) {
+        this->counter--;
+    }
+    if (this->dyna.actor.world.rot.y == 0) {
+        posOffset = 24.0f;
+    } else {
+        posOffset = -24.0f;
+    }
```
Comment (petrie911) [inline src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c:178]: possible ternary?

File: src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c [hunk 7]
```diff
@@ -31,25 +37,195 @@ const ActorInit Bg_Haka_Huta_InitVars = {
...
+    if (this->counter == 0) {
+        func_8087D268(this, globalCtx, NA_SE_EV_COFFIN_CAP_BOUND);
+        this->actionFunc = func_8087D720;
+    }
+}
+
+void func_8087D720(BgHakaHuta* this, GlobalContext* globalCtx) {
+    static Vec3f D_8087D958 = { 30.0f, 0.0f, 0.0f };
+    static Vec3f D_8087D964 = { 0.032579999f, 0.325800001f, -0.944899976f };
```
Comment (petrie911) [inline src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c:190]: I'm not really sure what's going on here, but this happens to be a unit vector.

File: src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c [hunk 8]
```diff
@@ -31,25 +37,195 @@ const ActorInit Bg_Haka_Huta_InitVars = {
...
+        this->counter = 6;
+        this->actionFunc = func_8087D8C0;
+    }
+
+    D_8087D958.x = this->counter + 24.0f;
+    if (D_8087D958.x > 30.0f) {
+        D_8087D958.x = 30.0f;
+    }
+    Matrix_RotateY(this->dyna.actor.world.rot.y * (M_PI / 32768), 0);
+    func_800D23FC(this->counter * (191 * M_PI / 3750), &D_8087D964, 1);
```
Comment (petrie911) [inline src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c:212]: What an oddly specific fraction.

File: src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c [hunk 9]
```diff
@@ -31,25 +37,195 @@ const ActorInit Bg_Haka_Huta_InitVars = {
...
+    } else {
+        pos.z = this->dyna.actor.world.pos.z - 120.0f;
+    }
+    pos.x = this->dyna.actor.world.pos.x;
+    pos.y = this->dyna.actor.world.pos.y;
+    Audio_PlaySoundAtPosition(globalCtx, &pos, 30, sfx);
+}
+
+void BgHakaHuta_SpawnEnemies(BgHakaHuta* this, GlobalContext* globalCtx) {
+    if ((Flags_GetSwitch(globalCtx, this->dyna.actor.params) != 0) && (Player_InCsMode(globalCtx) == 0)) {
```
Comment (petrie911) [inline src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c:116]:
```suggestion
    if (Flags_GetSwitch(globalCtx, this->dyna.actor.params) && !Player_InCsMode(globalCtx)) {
```

File: src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c [hunk 10]
```diff
@@ -31,25 +37,195 @@ const ActorInit Bg_Haka_Huta_InitVars = {
...
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Haka_Huta/BgHakaHuta_Draw.s")
+void BgHakaHuta_Init(Actor* thisx, GlobalContext* globalCtx) {
+    BgHakaHuta* this = THIS;
+    DynaPolyActor* new_var;
+    CollisionHeader* colHeader;
+
+    // Likely fake, but only way to make a match
+    new_var = &this->dyna;
+    colHeader = 0;
```
Comment (petrie911) [inline src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c:52]:
```suggestion
    colHeader = NULL;
```

File: src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c (hunk 1 above)
Comment (fig02) [reply src/overlays/actors/ovl_Bg_Haka_Huta/z_bg_haka_huta.c:51]: thisx should def achieve the same thing yea

**Review by Roman971 — APPROVED** [review]

**Review by fig02 — APPROVED** [review]

**Pull request merged**